client/blocks/upload-image/index.jsx

```diff
@@ -0,0 +1,199 @@
+
+		if ( ALLOWED_FILE_EXTENSIONS.indexOf( extension ) === -1 ) {
+			return;
+		}
+
+		const imageObjectUrl = URL.createObjectURL( files[ 0 ] );
+
+		this.setState( {
+			isEditingImage: true,
+			uploadedImage: imageObjectUrl
+		} );
+	};
+
+	onImageEditorDone = ( error, imageBlob ) => {
+		this.hideImageEditor();
+
+		if ( error ) {
+			return;
+		}
+
+		this.setState( { editedImage: URL.createObjectURL( imageBlob ) } );
+
+		this.props.onImageEditorDone( imageBlob );
+	};
+
+	hideImageEditor = () => {
+		const {
+			resetAllImageEditorState: resetAllImageEditorStateAction
+		} = this.props;
+
+		resetAllImageEditorStateAction();
+
+		URL.revokeObjectURL( this.state.uploadedImage );
+
+		this.setState( {
+			isEditingImage: false,
+			uploadedImage: false
+		} );
+	};
+
+	renderImageEditor() {
+		const {
+			additionalImageEditorClasses,
+			imageEditorProps,
+			texts,
+		} = this.props;
+
+		const {
+			isEditingImage,
+			uploadedImage
+		} = this.state;
+
+		const classes = classnames( 'upload-image-modal', additionalImageEditorClasses );
+
+		if ( isEditingImage ) {
+			return (
+				<Dialog
+					additionalClassNames={ classes }
+					isVisible={ true }
+				>
+					<ImageEditor
+						{ ...imageEditorProps }
+						media={ { src: uploadedImage } }
+						onDone={ this.onImageEditorDone }
+						onCancel={ this.hideImageEditor }
+						doneButtonText={ texts.doneButtonText ? texts.doneButtonText : 'Done' }
+					/>
+				</Dialog>
+			);
+		}
+	}
+
+	render() {
+		const {
+			children,
+			isUploading,
+			translate,
+			texts,
+			additionalClasses
+		} = this.props;
+
+		let {
+			placeholderContent,
+			uploadingContent,
+		} = this.props;
+
+		if ( typeof placeholderContent === 'undefined' ) {
+			placeholderContent = (
+				<div className="upload-image__placeholder">
+						<Gridicon icon="add-image" size={ 36 } />
+					<span>Add an image</span>
+				</div>
+			);
+		}
+
+		if ( typeof uploadingContent === 'undefined' ) {
+			const { editedImage } = this.state;
+
+			uploadingContent = (
+				<div className="upload-image__uploading-container">
+					<img src={ editedImage } />
+					<Spinner className="upload-image__spinner" size={ 40 } />
+				</div>
+			);
+		}
+
+		return (
+			<div className={ classnames( 'upload-image', additionalClasses ) } >
+				{ this.renderImageEditor() }
+
+				<FilePicker accept="image/*" onPick={ this.onReceiveFile }>
+					<div
+						className={
+							classnames( 'upload-image__image-container',
+								{ 'is-uploading': isUploading }
+							)
+						}
+					>
+						<DropZone
+							textLabel={ texts.dragUploadText
+								? texts.dragUploadText
+								: translate( 'Drop to upload image' )
+							}
+							onFilesDrop={ this.onReceiveFile }
+						/>
+
+						{ children }
+
+						{ ! isUploading && placeholderContent }
+						{ isUploading && uploadingContent }
+					</div>
+				</FilePicker>
+			</div>
+		);
+	}
+}
+
+export default connect(
+	null,
+	{
+		resetAllImageEditorState,
```

Comment:
Haha pretty strong there, but we were looking into having the editor be more encapsulated, so we have a bit more flexibility with the editor's render loop. Provided that we still have some way to talk to the editor in a reasonable way, there's no reason why we wouldn't be able to move away from having all state in redux.

See conversation in https://github.com/Automattic/wp-calypso/issues/6642#issuecomment-312516690